src/box/txn_limbo.c

```diff
@@ -1086,6 +1086,36 @@ txn_limbo_filter_request(struct txn_limbo *limbo,
 	}
 }
 
+/**
+ * Update the state of synchronous replication for system spaces.
+ */
+static void
+txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
+					     const struct synchro_request *req,
+					     bool is_rollback)
+{
+	/* Do not enable synchronous replication during bootstrap. */
+	if (req->origin_id == REPLICA_ID_NIL)
+		return;
+	uint16_t req_type = req->type;
+	assert(req_type == IPROTO_RAFT_PROMOTE ||
+	       req_type == IPROTO_RAFT_DEMOTE);
+	if (is_rollback) {
+		/* Reverse the request type. */
+		req_type = req_type == IPROTO_RAFT_PROMOTE ?
+			   IPROTO_RAFT_DEMOTE : IPROTO_RAFT_PROMOTE;
+	}
+	/* Synchronous replication is already enabled. */
+	if (!is_rollback && req_type == IPROTO_RAFT_PROMOTE &&
+	    limbo->owner_id != REPLICA_ID_NIL)
+		return;
+	/* Synchronous replication is already disabled. */
+	if (!is_rollback && req_type == IPROTO_RAFT_DEMOTE &&
+	    limbo->owner_id == REPLICA_ID_NIL)
+		return;
```

Comment:
This place seems incorrect.
Looks like in case you already have `limbo->owner_id != 0` set by some previous PROMOTE, and now another PROMOTE request is rolled back, you will make system spaces asynchronous. If I'm correct, please cover this with an extra test.

Let's take a look at all the 4 cases. Note that during a rollback `owner_id` is already in the 'old' state, because it's only changed inside `txn_limbo_req_commit()`.
1. Normal PROMOTE: if `owner_id != 0`, do nothing (because PROMOTE always sets a non-zero id, and we already have `owner_id != 0`)
2. Rollback of PROMOTE: if owner_id != 0, do nothing (because we are rolling back a PROMOTE, and during a corresponding `req_prepare` we did nothing. See point 1)
3. Normal DEMOTE: if `owner_id == 0`, do nothing (because DEMOTE always sets a zero id, and we already have `owner_id == 0`)
4. Rollback of DEMOTE: if `owner_id == 0`, do nothing (because we are rolling back a DEMOTE, and during a corresponding `req_prepare` we did nothing. See point 3)

Long story short, I'd rewrite this somewhat along these lines:
```diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index 61444ae84c..533fe08d0b 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -1100,20 +1100,17 @@ txn_limbo_update_system_spaces_is_sync_state(struct txn_limbo *limbo,
        uint16_t req_type = req->type;
        assert(req_type == IPROTO_RAFT_PROMOTE ||
               req_type == IPROTO_RAFT_DEMOTE);
-       if (is_rollback) {
-               /* Reverse the request type. */
-               req_type = req_type == IPROTO_RAFT_PROMOTE ?
-                          IPROTO_RAFT_DEMOTE : IPROTO_RAFT_PROMOTE;
-       }
+       bool is_promote = req_type == IPROTO_RAFT_PROMOTE;
        /* Synchronous replication is already enabled. */
-       if (!is_rollback && req_type == IPROTO_RAFT_PROMOTE &&
-           limbo->owner_id != REPLICA_ID_NIL)
+       if (is_promote && limbo->owner_id != REPLICA_ID_NIL)
                return;
        /* Synchronous replication is already disabled. */
-       if (!is_rollback && req_type == IPROTO_RAFT_DEMOTE &&
-           limbo->owner_id == REPLICA_ID_NIL)
+       if (!is_promote && limbo->owner_id == REPLICA_ID_NIL)
                return;
-       system_spaces_update_is_sync_state(req_type == IPROTO_RAFT_PROMOTE);
+       /* Flip operation types for a rollback. */
+       if (is_rollback)
+               is_promote = !is_promote;
+       system_spaces_update_is_sync_state(is_promote);
 }
 
 int
```

Reply:
Thanks for the catch!

> Note that during a rollback owner_id is already in the 'old' state, because it's only changed inside txn_limbo_req_commit().

Indeed, this is my misconception, I thought that the state is changed earlier.

> Long story short, I'd rewrite this somewhat along these lines:

Applied the changes you suggested. 

> Rollback of PROMOTE: if owner_id != 0, do nothing (because we are rolling back a PROMOTE, and during a corresponding req_prepare we did nothing. See point 1)

Added a test case for this.

> Rollback of DEMOTE: if owner_id == 0, do nothing (because we are rolling back a DEMOTE, and during a corresponding req_prepare we did nothing. See point 3)

Seems like this case is impossible:
https://github.com/tarantool/tarantool/blob/62c49367a2eab6d8008da2d7201ffa03e002c31e/src/box/box.cc#L3120-L3121

Added an assertion for it.